substrate/frame/staking/src/pallet/impls.rs

```diff
@@ -687,47 +705,118 @@ impl<T: Config> Pallet<T> {
+	) -> Result<(), ()> {
+		// preparing the next era. Note: we expect `do_elect_paged` to be called *only* during a
+		// non-genesis era, thus current era should be set by now.
+		let planning_era = CurrentEra::<T>::get().defensive_unwrap_or_default().saturating_add(1);
+
+		match Self::add_electables(supports.iter().map(|(s, _)| s.clone())) {
+			Ok(_) => {
+				let _ = Self::store_stakers_info(
+					Self::collect_exposures(supports).into_inner(),
+					planning_era,
+				);
+				Ok(())
+			},
+			Err(not_included) => {
+				log!(
+					warn,
+					"not all winners fit within the electable stashes, excluding tail: {:?}.",
+					not_included
+				);
+
+				// filter out supports of stashes that do not fit within the electable stashes
+				// storage bounds to prevent collecting their exposures.
+				supports.retain(|(s, _)| !not_included.contains(s));
+
+				let _ = Self::store_stakers_info(
+					Self::collect_exposures(supports).into_inner(),
+					planning_era,
+				);
+				Err(())
+			},
+		}
+	}
+
+	/// Process the output of a paged election.
+	///
+	/// Store staking information for the new planned era of a single election page.
 	pub fn store_stakers_info(
-		exposures: BoundedVec<
-			(T::AccountId, Exposure<T::AccountId, BalanceOf<T>>),
-			MaxWinnersOf<T>,
-		>,
+		exposures: Vec<(T::AccountId, Exposure<T::AccountId, BalanceOf<T>>)>,
 		new_planned_era: EraIndex,
-	) -> BoundedVec<T::AccountId, MaxWinnersOf<T>> {
-		// Populate elected stash, stakers, exposures, and the snapshot of validator prefs.
-		let mut total_stake: BalanceOf<T> = Zero::zero();
-		let mut elected_stashes = Vec::with_capacity(exposures.len());
+	) -> BoundedVec<T::AccountId, MaxWinnersPerPageOf<T::ElectionProvider>> {
+		// populate elected stash, stakers, exposures, and the snapshot of validator prefs.
+		let mut total_stake_page: BalanceOf<T> = Zero::zero();
+		let mut elected_stashes_page = Vec::with_capacity(exposures.len());
 
 		exposures.into_iter().for_each(|(stash, exposure)| {
-			// build elected stash
-			elected_stashes.push(stash.clone());
-			// accumulate total stake
-			total_stake = total_stake.saturating_add(exposure.total);
-			// store staker exposure for this era
-			EraInfo::<T>::set_exposure(new_planned_era, &stash, exposure);
+			// build elected stash.
+			elected_stashes_page.push(stash.clone());
+			// accumulate total stake.
+			total_stake_page = total_stake_page.saturating_add(exposure.total);
+			// set or update staker exposure for this era.
+			EraInfo::<T>::upsert_exposure(new_planned_era, &stash, exposure);
 		});
 
-		let elected_stashes: BoundedVec<_, MaxWinnersOf<T>> = elected_stashes
-			.try_into()
-			.expect("elected_stashes.len() always equal to exposures.len(); qed");
+		let elected_stashes: BoundedVec<_, MaxWinnersPerPageOf<T::ElectionProvider>> =
+			elected_stashes_page
+				.try_into()
+				.expect("elected_stashes.len() always equal to exposures.len(); qed");
```

Comment:
But in the API of the function the exposures come unbounded. From what I can see, they are always bounded at the call site but then unwrapped with `into_inner`. Maybe we should make this function take the bounded vec as a parameter.